File: doc/codes/python/examples/api/compute/holo/gspat.py
import numpy as np
from autd3.geometry import Autd3, Geometry
from autd3.units import m, s
from autd3.value import Intensity
from autd3_pattern import TransducerMask
from autd3_pattern import wavelength as calc_wavelength
from autd3_pattern_holo import (
    AmplitudeTarget,
    Directivity,
    IntensityConstraint,
    GspatOption,
    Pa,
    gspat,
)

geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

center = geometry.center() + np.array([0.0, 0.0, 150.0])
foci = [
    AmplitudeTarget(
        center + np.array([-30.0, 0.0, 0.0]),
        2.5e3 * Pa,
    ),
    AmplitudeTarget(
        center + np.array([30.0, 0.0, 0.0]),
        2.5e3 * Pa,
    ),
]

wavelength = calc_wavelength(340 * m / s)
repeat = 100
constraint = IntensityConstraint.Clamp(Intensity.MIN, Intensity.MAX)
directivity = Directivity.Sphere
mask = TransducerMask.AllEnabled
parallel = True
option = (
    # ANCHOR: option
    GspatOption(
        repeat,
        constraint,
        directivity,
        mask,
        parallel,
    )
    # ANCHOR_END: option
)
phases = geometry.phase_buffer()
intensities = geometry.intensity_buffer()
# ANCHOR: api
gspat(geometry, foci, wavelength, option, phases, intensities)
# ANCHOR_END: api

File: doc/codes/rust/examples/api/compute/holo/gspat.rs
use std::num::NonZeroUsize;

use anyhow::Result;

use autd3_rs::geometry::{Autd3, Geometry, TransducerMask, offset};
use autd3_rs::units::{m, mm, s};
use autd3_rs::value::Intensity;
use autd3_rs_pattern_holo::{
    AmplitudeTarget, Directivity, IntensityConstraint, GspatOption, NalgebraBackend, Pa, gspat,
};

fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let center = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let foci = [
        AmplitudeTarget {
            point: center + offset(-30.0 * mm, 0.0 * mm, 0.0 * mm),
            amplitude: 2.5e3 * Pa,
        },
        AmplitudeTarget {
            point: center + offset(30.0 * mm, 0.0 * mm, 0.0 * mm),
            amplitude: 2.5e3 * Pa,
        },
    ];

    let wavelength = autd3_rs_pattern::wavelength(340.0 * m / s);
    let repeat = NonZeroUsize::new(100).unwrap();
    let constraint = IntensityConstraint::Clamp(Intensity::MIN, Intensity::MAX);
    let directivity = Directivity::Sphere;
    let mask = TransducerMask::AllEnabled;
    let parallel = true;
    let option =
        // ANCHOR: option
        GspatOption {
            repeat,
            constraint,
            directivity,
            mask,
            parallel,
            ..Default::default()
        }
        // ANCHOR_END: option
        ;
    let mut phases = geometry.phase_buffer();
    let mut intensities = geometry.intensity_buffer();
    // ANCHOR: api
    gspat(
        &NalgebraBackend,
        &geometry,
        &foci,
        wavelength,
        &option,
        &mut phases,
        &mut intensities,
    )?;
    // ANCHOR_END: api
    Ok(())
}
